src/high-level/constructors.lisp

```diff
@@ -126,31 +126,36 @@ The tensor is specialized on SHAPE and TYPE."
       ((type shape shape))
     (let* ((tensor-class (infer-tensor-type type shape nil))
            (storage-size (reduce #'* shape))
-           (storage (make-array storage-size :element-type type))
            (array-dims (array-dimensions array)))
-      (let ((index-function
-              (if (eq layout ':row-major)
-                  #'row-major-index
-                  #'column-major-index))
-            (input-index-function
-              (if (eq input-layout ':row-major)
-                  #'row-major-index
-                  #'column-major-index)))
-        (cond
-          ((not (cdr array-dims))
-           (map-indexes
-            shape
-            (lambda (&rest pos)
-              (setf (aref storage (funcall index-function pos shape))
-                    (aref array (funcall input-index-function pos shape))))))
-          (t (map-indexes
+      (multiple-value-bind (storage finalizer)
+          (allocate storage-size
```

Comment:
Sorry to bug this PR. I think the recent Franz port told us that not specifying an `initial-element` is undefined behavior.

Maybe we default the `element-type` to `*default-tensor-type*` (defined [here](https://github.com/quil-lang/magicl/blob/master/src/high-level/constructors.lisp#L7)), and default `initial-element` to `(coerce 0 element-type)` so we don't accidentally encounter this issue?

Reply:
I hadn't considered the discussion in that PR, but afaict not specifing `:initial-element` seems to be perfectly fine for constructing the array -- the undefined behavior is if we read before it's set, http://clhs.lisp.se/Body/f_mk_ar.htm

```
If initial-element is not supplied, the consequences of later reading an uninitialized element of new-array are undefined unless either initial-contents is supplied or displaced-to is non-nil.
```

So this is just the `malloc` rather than `calloc` behavior. I'm fine with zero initializing, but I don't see why it is required (and I wouldn't want anybody to come to expect it -- that's what the `zeros` constructor is for).

Reply:
(that CLHS page even gives an example in the code listing at the bottom)

```
 (make-array 5 :element-type 'single-float) ;; Creates an array of single-floats.
```